### PR TITLE
eCDN live changes

### PR DESCRIPTION
**Pull request by yossipapi**

### CODE REVIEW

File: api_v3/services/ServerNodeService.php [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ * @subpackage services
+ */
+class ServerNodeService extends KalturaBaseService
+{
+	public function initService($serviceId, $serviceName, $actionName)
+	{
+		parent::initService($serviceId, $serviceName, $actionName);
+		
+		$partnerId = $this->getPartnerId();
+		if(!$this->getPartner()->getEnabledService(PermissionName::FEATURE_SERVER_NODE))
```
Comment (tan-tan-kanarek) [inline api_v3/services/ServerNodeService.php]: If all actions of the service already protected using the API roles and permissions, why do you add that check here hard-coded?

File: api_v3/lib/types/serverNode/base/KalturaServerNode.php [hunk 2]
```diff
@@ -49,51 +56,52 @@ class KalturaEdgeServer extends KalturaObject implements IFilterable
...
 	/**
-	 * edgeServer host name
-	 * 
-	 * @var string
-	 * @filter like,mlikeor,mlikeand
+	 * @var KalturaServerNodeStatus
+	 * @filter eq,in
 	 */
-	public $hostName;
+	public $status;
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/base/KalturaServerNode.php]: Please make the status read only, is should be set using enable and disable actions.

File: api_v3/lib/types/serverNode/base/KalturaServerNode.php [hunk 3]
```diff
@@ -164,55 +160,52 @@ public function validateMandatoryAttributes($isInsert = false)
...
-		if(is_null($object_to_fill))
-			$object_to_fill = new EdgeServer();
-		
-		
-		$object_to_fill =  parent::toObject($object_to_fill, $props_to_skip);
+		parent::doFromObject($source_object, $responseProfile);
 		
-		return $object_to_fill;
+		if($source_object->getHeartbeatTime() < (time() - 90))
+			$this->status = ServerNodeStatus::NOT_REGISTERED;
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/base/KalturaServerNode.php]: I think that it shouldn't return NOT_REGISTERED if it's DISABLED.

File: alpha/lib/model/ServerNode.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+	/**
+	 * Initializes internal state of ServerNode object.
+	 * @see        parent::__construct()
+	 */
+	public function __construct()
+	{
+		// Make sure that parent constructor is always invoked, since that
+		// is where any default values for this object are set.
+		parent::__construct();
```
Comment (tan-tan-kanarek) [inline alpha/lib/model/ServerNode.php]: The parent constructor will be called anyway, this empty method is redundant.

File: plugins/media/wowza/services/LiveConversionProfileService.php [hunk 5]
```diff
@@ -61,7 +61,7 @@ public function serveAction($streamName, $hostname = null)
 			
 		$mediaServer = null;
 		if($hostname)
-			$mediaServer = MediaServerPeer::retrieveByHostname($hostname);
+			$mediaServer = ServerNodePeer::retrieveByHostname($hostname);
```
Comment (tan-tan-kanarek) [inline plugins/media/wowza/services/LiveConversionProfileService.php]: Validate that $mediaServer is not disabled.

File: alpha/lib/model/LiveEntry.php [hunk 6]
```diff
@@ -636,11 +617,9 @@ public function setMediaServer($index, $hostname, $applicationName = null)
 		if(is_null($this->getFirstBroadcast())) 
 			$this->setFirstBroadcast(kApiCache::getTime());
 				
-		$mediaServer = MediaServerPeer::retrieveByHostname($hostname);
-		if (!$mediaServer)
-		{
-			KalturaLog::info("External media server with hostname [$hostname] is being used to stream this entry");
-		}
+		$mediaServerNode = ServerNodePeer::retrieveByHostName($hostname);
+		if (!$mediaServerNode)
```
Comment (tan-tan-kanarek) [inline alpha/lib/model/LiveEntry.php]: Validate that $mediaServerNode is not disabled.

File: api_v3/services/ServerNodeService.php [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+		{
+			if($serverNode)
+			{
+				return $this->addAction($serverNode);
+			}
+			else 
+				throw new KalturaAPIException(KalturaErrors::SERVER_NODE_NOT_FOUND, $hostName);
+		}
+	
+		$dbServerNode->setHeartbeatTime(time());
```
Comment (tan-tan-kanarek) [inline api_v3/services/ServerNodeService.php]: Validate that $dbServerNode is not disabled.

File: api_v3/services/ServerNodeService.php [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	function reportStatusAction($hostName, KalturaServerNode $serverNode = null)
+	{
+		$dbServerNode = ServerNodePeer::retrieveByPartnerIdAndHostName($this->getPartnerId(), $hostName);
+		
+		//Allow serverNodes auto registration without calling add
+		if (!$dbServerNode)
+		{
+			if($serverNode)
+			{
+				return $this->addAction($serverNode);
```
Comment (tan-tan-kanarek) [inline api_v3/services/ServerNodeService.php]: addAction doesn't set the heartbeat.

File: api_v3/services/ServerNodeService.php (hunk 7 above)
Comment (tan-tan-kanarek) [inline api_v3/services/ServerNodeService.php]: I think you should also call toUpdateable in order to apply status attributes from $serverNode to $dbServerNode.

File: api_v3/lib/KalturaEntryService.php [hunk 9]
```diff
@@ -322,7 +322,7 @@ protected function attachLiveEntryResource(kLiveEntryResource $resource, entry $
 			if(!$mediaServer)
 				throw new KalturaAPIException(KalturaErrors::NO_MEDIA_SERVER_FOUND, $dbLiveEntry->getId());
 				
-			$mediaServerLiveService = $mediaServer->getWebService(MediaServer::WEB_SERVICE_LIVE);
+			$mediaServerLiveService = $mediaServer->getWebService($mediaServer->getLiveWebServiceIdentifier());
```
Comment (tan-tan-kanarek) [inline api_v3/lib/KalturaEntryService.php]: Where is getLiveWebServiceIdentifier implemented?

File: api_v3/lib/types/serverNode/base/KalturaServerNode.php [hunk 10]
```diff
@@ -3,7 +3,7 @@
  * @package api
  * @subpackage objects
  */
-class KalturaEdgeServer extends KalturaObject implements IFilterable
+abstract class KalturaServerNode extends KalturaObject implements IRelatedFilterable, IApiObjectFactory
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/base/KalturaServerNode.php]: Why IRelatedFilterable?

File: api_v3/lib/types/serverNode/base/KalturaServerNode.php [hunk 11]
```diff
@@ -164,55 +160,52 @@ public function validateMandatoryAttributes($isInsert = false)
...
 		
-		if($edgeId)
-			$c->add(EdgeServerPeer::ID, $edgeId, Criteria::NOT_EQUAL);
+		if($serverNodeId)
+			$c->add(ServerNodePeer::ID, $serverNodeId, Criteria::NOT_EQUAL);
 		
-		$c->add(EdgeServerPeer::HOST_NAME, $this->hostName);
-		$c->add(EdgeServerPeer::STATUS, array(EdgeServerStatus::ACTIVE, EdgeServerStatus::DISABLED), Criteria::IN);
+		$c->add(ServerNodePeer::HOST_NAME, $this->hostName);
+		$c->add(ServerNodePeer::STATUS, array(ServerNodeStatus::ACTIVE, ServerNodeStatus::DISABLED), Criteria::IN);
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/base/KalturaServerNode.php]: No need to add criteria on status as the default criteria already does it.

File: api_v3/lib/types/serverNode/base/KalturaServerNode.php [hunk 12]
```diff
@@ -164,55 +160,52 @@ public function validateMandatoryAttributes($isInsert = false)
...
 	{
-		if(is_null($object_to_fill))
-			$object_to_fill = new EdgeServer();
-		
-		
-		$object_to_fill =  parent::toObject($object_to_fill, $props_to_skip);
+		parent::doFromObject($source_object, $responseProfile);
 		
-		return $object_to_fill;
+		if($source_object->getHeartbeatTime() < (time() - 90))
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/base/KalturaServerNode.php]: Check if $this->shouldGet('status', $responseProfile).

File: api_v3/lib/types/serverNode/base/KalturaServerNode.php [hunk 13]
```diff
@@ -230,5 +223,14 @@ public function getFilterDocs()
 	{
 		return array();
 	}
-}
-
+	
+	public static function getInstance($sourceObject, KalturaDetachedResponseProfile $responseProfile = null)
+	{
+		$object = KalturaServerNodeFactory::getInstanceByType($sourceObject->getType());
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/base/KalturaServerNode.php]: Why don't you implement KalturaServerNodeFactory::getInstanceByType here?

File: api_v3/lib/types/serverNode/base/KalturaServerNodeArray.php [hunk 14]
```diff
@@ -3,15 +3,15 @@
...
 {
 	public static function fromDbArray(array $arr, KalturaDetachedResponseProfile $responseProfile = null)
 	{
-		$newArr = new KalturaEdgeServerArray();
+		$newArr = new KalturaServerNodeArray();
 		foreach($arr as $obj)
 		{
 		    /* @var $obj StorageProfile */
-			$nObj = new KalturaEdgeServer();
+			$nObj = KalturaServerNodeFactory::getInstanceByType($obj->getType());
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/base/KalturaServerNodeArray.php]: Call KalturaServerNode::getInstance instead.

File: api_v3/services/ServerNodeService.php [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+	function addAction(KalturaServerNode $serverNode)
+	{	
+		if(!$serverNode->status)
+			$serverNode->status = KalturaServerNodeStatus::DISABLED; 
+		
+		$dbServerNode = $serverNode->toInsertableObject();
+		$dbServerNode->setPartnerId($this->getPartnerId());
+		$dbServerNode->save();
+		
+		$serverNode = KalturaServerNodeFactory::getInstanceByType($dbServerNode->getType());
```
Comment (tan-tan-kanarek) [inline api_v3/services/ServerNodeService.php]: Call KalturaServerNode::getInstance instead.

File: api_v3/services/ServerNodeService.php [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+				return $this->addAction($serverNode);
+			}
+			else 
+				throw new KalturaAPIException(KalturaErrors::SERVER_NODE_NOT_FOUND, $hostName);
+		}
+	
+		$dbServerNode->setHeartbeatTime(time());
+		$dbServerNode->save();
+	
+		$serverNode = KalturaServerNodeFactory::getInstanceByType($dbServerNode->getType());
```
Comment (tan-tan-kanarek) [inline api_v3/services/ServerNodeService.php]: Call KalturaServerNode::getInstance instead.

File: api_v3/lib/types/serverNode/base/filters/KalturaServerNodeFilter.php [hunk 17]
```diff
@@ -0,0 +1,49 @@
+<?php
+/**
+ * @package api
+ * @subpackage filters
+ */
+class KalturaServerNodeFilter extends KalturaServerNodeBaseFilter
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/base/filters/KalturaServerNodeFilter.php]: There is no reason to extend and implement KalturaRelatedFilter.

File: api_v3/lib/types/serverNode/deliveryServer/mediaServerNode/KalturaMediaServerNode.php [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+	/* (non-PHPdoc)
+	 * @see KalturaObject::toObject()
+	 */
+	public function toObject($dbObject = null, $skip = array())
+	{
+		$dbObject = parent::toObject($dbObject, $skip);
+	
+		if (!is_null($this->protocolPortConfig))
+		{
+			$mediaServerPortConfig = array();
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/deliveryServer/mediaServerNode/KalturaMediaServerNode.php]: Just call $dbObject->setMediaServerPortConfig($this->protocolPortConfig->toObjectsArray());

File: api_v3/lib/types/serverNode/deliveryServer/mediaServerNode/KalturaMediaServerNode.php [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+	
+	/* (non-PHPdoc)
+	 * @see KalturaObject::fromObject()
+	 */
+	public function doFromObject($source_object, KalturaDetachedResponseProfile $responseProfile = null)
+	{
+		/* @var $source_object MediaServerNode */
+		parent::doFromObject($source_object, $responseProfile);
+	
+		if($this->shouldGet('mediaServerPortConfig', $responseProfile) && !is_null($source_object->getMediaServerPortConfig()))
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/deliveryServer/mediaServerNode/KalturaMediaServerNode.php]: KalturaObject already calls fromDbArray which calls fromKeyValueArray, so you don't need any special treatment.

File: api_v3/lib/types/serverNode/base/KalturaServerNodeArray.php [hunk 20]
```diff
@@ -3,15 +3,15 @@
...
 {
 	public static function fromDbArray(array $arr, KalturaDetachedResponseProfile $responseProfile = null)
 	{
-		$newArr = new KalturaEdgeServerArray();
+		$newArr = new KalturaServerNodeArray();
 		foreach($arr as $obj)
 		{
 		    /* @var $obj StorageProfile */
-			$nObj = new KalturaEdgeServer();
+			$nObj = KalturaServerNode::getInstance($obj);
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/base/KalturaServerNodeArray.php]: KalturaServerNode::getInstance expecting also to get the $responseProfile, and fromObject is not required after it.

Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/base/KalturaServerNodeArray.php]: Same for all calls to KalturaServerNode::getInstance.

File: api_v3/lib/types/serverNode/deliveryServer/mediaServerNode/KalturaMediaServerNode.php [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+	 */
+	public function getMapBetweenObjects()
+	{
+		return array_merge(parent::getMapBetweenObjects(), self::$mapBetweenObjects);
+	}
+	
+	/* (non-PHPdoc)
+	 * @see KalturaObject::toObject()
+	 */
+	public function toObject($dbObject = null, $skip = array())
```
Comment (tan-tan-kanarek) [inline api_v3/lib/types/serverNode/deliveryServer/mediaServerNode/KalturaMediaServerNode.php]: The entire toObject method could be removed, the KalturaObject already calls toObjectsArray for you.

File: api_v3/services/ServerNodeService.php [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+	}
+	
+	/**
+	 * delete server node by id
+	 *
+	 * @action delete
+	 * @param string $serverNodeId
+	 * @throws KalturaErrors::INVALID_OBJECT_ID
+	 */
+	function deleteAction($serverNodeId)
```
Comment (tan-tan-kanarek) [inline api_v3/services/ServerNodeService.php]: Enable and disable actions should return a KalturaServerNode object.

Comment (yossipapi) [inline api_v3/services/ServerNodeService.php]: Done

Comment (yossipapi) [inline api_v3/services/ServerNodeService.php]: @tan-tan-kanarek done with the changes that were required.

**Pull request merged**